src/utils.cpp

```diff
@@ -334,6 +335,621 @@ void cnf_to_dimacs(Term cnf, std::ostringstream & y)
   }
 }
 
+// mapping each subformula with a new name and returning a vector of pair of
+// terms. Each pair consists of the parent term and it's children(for each
+// subformula)
+class TseitinTraversal : public IdentityWalker
+{
+ public:
+  std::vector<std::pair<Term, Term>>
+      reduced;  // stores a pair of (lhs, rhs) in x1(lhs)<->(formula(rhs))
+
+  TseitinTraversal(SmtSolver solver_) : IdentityWalker{ solver_, false } {}
+  WalkerStepResult visit_term(Term & term)
+  {
+    Sort boolsort = term->get_sort();
+    assert(term->get_sort() == boolsort);
+
+    auto give_symbolic_name = [&](Term t) {  // producing a new symbol
+      int pt = 1;
+      while (true)
+      {
+        try
+        {
+          return solver_->make_symbol("tseitin_to_cnf_" + std::to_string(pt),
+                                      boolsort);
+        }
+        catch (IncorrectUsageException & e)
+        {
+          pt++;
+        }
+      }
+
+    };
+    if (!preorder_)  // using post order traversal as I need the new names of
+                     // the children to generate the new term
+    {
+      smt::Op op = term->get_op();
+      if (!op.is_null())
+      {
+        std::vector<Term> vec;  // a vector of all children
+        for (auto u : term)
+        {
+          Term term_name;
+          bool present = query_cache(
+              u,
+              term_name);  // finding the new name of each child from the cache
+          assert(present == true);
+          vec.push_back(term_name);
+        }
+        Term term_name;
```

Comment:
It is a bit confusing to have again this declaration. If it is conceptually different, I suggest using a different name.